crates/ruff/src/rules/pycodestyle/logical_lines.rs

```diff
@@ -25,78 +28,327 @@ bitflags! {
-) -> LogicalLine<'a> {
-    let mut logical = String::with_capacity(88);
-    let mut mapping = Vec::new();
-    let mut flags = TokenFlags::empty();
-    let mut prev: Option<&Location> = None;
-    let mut length = 0;
-    for (start, tok, end) in &tokens {
+    mappings: Mappings,
+    lines: Vec<Line>,
+    current_line: Option<CurrentLine>,
+}
+
+impl<'a> LogicalLinesBuilder<'a> {
+    fn with_token_capacity(capacity: usize) -> Self {
+        Self {
+            tokens: Vec::with_capacity(capacity),
+            mappings: Mappings::with_capacity(capacity + 1),
+            ..Self::default()
+        }
+    }
+
+    // SAFETY: `LogicalLines::from_tokens` asserts that the file has less than `u32::MAX` tokens and each tokens is at least one character long
+    #[allow(clippy::cast_possible_truncation)]
+    fn push_token(&mut self, start: Location, token: &'a Tok, end: Location, locator: &Locator) {
+        let tokens_start = self.tokens.len();
+        self.tokens.push((start, token, end));
+
+        let mut line = self.current_line.get_or_insert_with(|| {
+            let mappings_start = self.mappings.len();
+            self.mappings.push(self.text.len(), start);
+
+            CurrentLine {
+                flags: TokenFlags::empty(),
+                text_start: self.text.len() as u32,
+                mappings_start: mappings_start as u32,
+                tokens_start: tokens_start as u32,
+                previous_token: None,
+            }
+        });
+
         if matches!(
-            tok,
+            token,
             Tok::Newline | Tok::NonLogicalNewline | Tok::Indent | Tok::Dedent
         ) {
-            continue;
+            return;
         }
 
-        if mapping.is_empty() {
-            mapping.push((0, *start));
+        if matches!(token, Tok::Comment(..)) {
+            line.flags.insert(TokenFlags::COMMENT);
+            return;
         }
 
-        if matches!(tok, Tok::Comment { .. }) {
-            flags.insert(TokenFlags::COMMENT);
-            continue;
-        }
-
-        if is_op_token(tok) {
-            flags.insert(TokenFlags::OPERATOR);
+        if is_op_token(token) {
+            line.flags.insert(TokenFlags::OPERATOR);
         }
 
         if matches!(
-            tok,
+            token,
             Tok::Lpar | Tok::Lsqb | Tok::Lbrace | Tok::Rpar | Tok::Rsqb | Tok::Rbrace
         ) {
-            flags.insert(TokenFlags::BRACKET);
+            line.flags.insert(TokenFlags::BRACKET);
         }
 
-        if matches!(tok, Tok::Comma | Tok::Semi | Tok::Colon) {
-            flags.insert(TokenFlags::PUNCTUATION);
+        if matches!(token, Tok::Comma | Tok::Semi | Tok::Colon) {
+            line.flags.insert(TokenFlags::PUNCTUATION);
         }
 
-        if is_keyword_token(tok) {
-            flags.insert(TokenFlags::KEYWORD);
+        if is_keyword_token(token) {
+            line.flags.insert(TokenFlags::KEYWORD);
         }
 
         // TODO(charlie): "Mute" strings.
```

Comment:
I don't know if this TODO is still relevant, it looks like we _are_ "muting" strings.

Reply:
It is still necessary (at least for this PR). My understanding was that this is a TODO to avoid the muting :D